
---
title: |
  ![](_css/logo.png){width=2in} SARS-CoV2 

date: "2020-06-13"
keywords: [Markdown, SARS-CoV19]
css:
- _css/pandoc.css
to: html5
from: markdown
...

# What information to find here 

## Database
[The directorty *data*](data/) contains the database generated by peroba using data from COGUK and GISAID (sequences and tables).

These files do not incorporate the newest information from our local master table and sequences. 
Below is a brief description of files on [directory *data*](data/) (only the suffixes, I'm not linking all since the 
names change a bit to include a timestamp...):

* `raw.csv.gz`: merged metadata from several COGUK csv files with distinct information,  and the GISAID metadata file 
* `metadata.csv.gz`: same as above, but only for samples for which we can find the equivalent sequence
* `sequences.fasta.xz`: unaligned sequences from COGUK and GISAID whose names can be found on the metadata
* `sequences.aln.xz`: reference-based alignment of the sequences above
* `tree.nhx`: phylogeny provided by COGUK pruned to leaves where a sequence was found
* `html`: a report from the `raw.csv.gz` file

Here is a [browseable description of the merged metadata](data/perobaDB.0608.html) (equivalent to the `raw.csv.gz` file).

## Report 
[The directory *report*](report/) contains the analysis using the database above with our sequences and master table
(metadata).

The [phylogenomics report can be found here](report/report_2020-06-09.html).

The two main files are the `csv` and the `metadata` tables, which contain respectively our local and global samples,
together with `peroba_` columns to designate phylogenetic imputation of lineages.  

## Custom analysis
[the directory *analysis*](analysis/) contains custom analysis to answer a specific question, or using more customised
scripts. Most analyses will be in Jupyter notebooks (with code). Please check the directory for the up-to-date files,
here is a short description:
[Maps for each *UK_lineage*- where a local sample was found](analysis/021.uk_lineage_maps.html)

## Disclaimer 
This website is for internal use only. 

The data compiled here comes from COGUK and GISAID, and while COGUK has an open data policy, GISAID is quite strict and
needs explicit consent. You may find COGUK data here before it becomes public, and all data has been processed 
by the bioinformatics team.

If you want to access or use this data in any way you **must** first contact [Leo Martins](mailto:Leonardo.de-Oliveira-Martins@quadram.ac.uk).


